[Clang] [Test] Use lit Syntax for Environment Variables in Clang subproject

Harini0924 opened the pull request:
This patch updates the clang tests by replacing shell command substitutions with lit-compatible syntax for setting and referencing environment variables. Specifically, the use of shell-style variable substitution (e.g., `DEFAULT_TRIPLE=`and `EXPECTED_RESOURCE_DIR=`) has been replaced with `env` and `%{env}` to align with lit's internal shell requirements. These changes ensure that environment variables are properly set and accessed within the lit environment.

When using the lit internal shell with the command `LIT_USE_INTERNAL_SHELL=1 ninja check-clang`, one common error encountered is:
```
FAIL: Clang :: Driver/program-path-priority.c (19 of 20640)
******************** TEST 'Clang :: Driver/program-path-priority.c' FAILED ********************
Exit Code: 127

Command Output (stdout):
--
# RUN: at line 90
DEFAULT_TRIPLE=`/usr/local/google/home/harinidonthula/llvm-project/build/tools/clang/test/Driver/Output/program-path-priority.c.tmp/clang --version | grep "Target:" | cut -d ' ' -f2`
# executed command: 'DEFAULT_TRIPLE=`/usr/local/google/home/harinidonthula/llvm-project/build/tools/clang/test/Driver/Output/program-path-priority.c.tmp/clang' --version
# .---command stderr------------
# | 'DEFAULT_TRIPLE=`/usr/local/google/home/harinidonthula/llvm-project/build/tools/clang/test/Driver/Output/program-path-priority.c.tmp/clang': command not found
# `-----------------------------
# error: command failed with exit status: 127
```
To fix this issue, the patch replaces traditional shell substitutions with lit's environment variable handling, ensuring compatibility with the lit internal shell framework. This update applies to both the handling of the `DEFAULT_TRIPLE` and `EXPECTED_RESOURCE_DIR` variables, allowing the tests to pass when using the lit internal shell.
The patch also adds `env` to the `PWD` variable setting in the following command to ensure the environment variable is correctly set within the lit internal shell:
```
// RUN: %if system-linux %{ env PWD=/proc/self/cwd %clang -### -c --coverage %s -o foo/bar.o 2>&1 | FileCheck --check-prefix=PWD %s %}
```
[link to RFC](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179)

Harini0924 committed:
[llvm-lit] Replace Shell Substitutions with lit Syntax for Environment Variables

This patch updates the llvm-lit clang tests by replacing shell command
substitutions with lit-compatible syntax for setting and referencing
environment variables. Specifically, the use of shell-style variable
substitution (e.g., DEFAULT_TRIPLE=\`` and EXPECTED_RESOURCE_DIR=`)
has been replaced with env and %{env}to align with lit's internal shell
requirements. The changes ensure that environment variables are properly
set and accessed within the lit environment. This update applies to both
the handling of the `DEFAULT_TRIPLE` variable `EXPECTED_RESOURCE_DIR` variable.
This change will allow the tests to pass and ensure the compatibilty
with the lit internal shell framework.

Harini0924 committed:
[llvm-lit] Update to Use lit Syntax for Environment Variables

This patch updates the llvm-lit clang tests by replacing shell
command substitutions with lit-compatible syntax for setting
and referencing environment variables. Shell-style variable
substitution (e.g., DEFAULT_TRIPLE=\`` and EXPECTED_RESOURCE_DIR=`)
has been replaced with env and%{env}to ensure compatibility with
lit's internal shell. This change fixes issues encountered when running
tests with LIT_USE_INTERNAL_SHELL=1 ninja check-clang, where shell
commands failed due to improper handling of environment variables.
Additionally, env has been added to the PWD variable setting to
ensure it is correctly established within the lit internal shell.
These updates ensure that the tests pass and function as expected
within the lit framework.

Harini0924 commented:
CC: @ilovepi @petrhosek

ilovepi commented on clang/test/Driver/program-path-priority.c:
```diff
@@ -87,12 +87,13 @@
+// RUN: env DEFAULT_TRIPLE=%{readfile:%t/default-triple.txt} 
+// RUN: touch %t/%{env:DEFAULT_TRIPLE}-gcc && chmod +x %t/%{env:DEFAULT_TRIPLE}-gcc 
+// RUN: touch %t/%target_triple-gcc && chmod +x %t/%target_triple-gcc && 
+// RUN: env "PATH=%t:$PATH" %t/clang -### -target notreal-none-elf %s 2>&1 | \
+// RUN: FileCheck --check-prefix=DEFAULT_TRIPLE_GCC %s
+// DEFAULT_TRIPLE_GCC: "gcc" "-o" "a.out"
```
Why are these lines different? I don't think they should be changed. At the very least, the check shouldn't have been updated.

ilovepi commented on clang/test/Driver/program-path-priority.c:
```diff
@@ -87,12 +87,13 @@
+// RUN: env DEFAULT_TRIPLE=%{readfile:%t/default-triple.txt} 
+// RUN: touch %t/%{env:DEFAULT_TRIPLE}-gcc && chmod +x %t/%{env:DEFAULT_TRIPLE}-gcc 
+// RUN: touch %t/%target_triple-gcc && chmod +x %t/%target_triple-gcc && 
+// RUN: env "PATH=%t:$PATH" %t/clang -### -target notreal-none-elf %s 2>&1 | \
```
Why is `PATH` set differently?

Harini0924 replied:
Thank you for catching this. You're absolutely right—the `PATH` setting and the check shouldn't have been changed. I was experimenting with a different implementation, and I mistakenly left those changes in before committing. However, I reverted these changes back to their original state in my latest patch, which restored the correct behavior.

Harini0924 committed:
[llvm-lit] Changed the PATH and FileCheck lines to original state

i accidentally changed the PATH and FileCheck lines in my last commit.
This patch changes it back to its original state

ilovepi commented on clang/test/Driver/program-path-priority.c:
```diff
@@ -87,8 +87,8 @@
 /// on PATH beats default triple in program path
-// RUN: DEFAULT_TRIPLE=`%t/clang --version | grep "Target:" | cut -d ' ' -f2`
-// RUN: touch %t/$DEFAULT_TRIPLE-gcc && chmod +x %t/$DEFAULT_TRIPLE-gcc
+// RUN: %clang --version | grep "Target:" | cut -d ' ' -f2 > %t/default-triple.txt
```
```suggestion
// RUN: %t/clang --version | grep "Target:" | cut -d ' ' -f2 > %t/default-triple.txt
```
Wasn't the test originally written that way? `%clang` is its own substitution that invokes the driver: https://llvm.org/docs/TestingGuide.html
While I don't think you'll get meaningfully different behavior, lets leave the existing behavior intact unless we have a reason to change it.

ilovepi reviewed (approved):
LGTM, once that last nit is fixed, assuming CI is passing.

Harini0924 committed:
[llvm-lit] added %t/clang
